pkg/seedadmissioncontroller/webhooks/admission/extensionresources/admission.go

```diff
@@ -0,0 +1,253 @@
+			}})
+
+	h.addResource(metav1.GroupVersionResource{Group: extensionsv1alpha1.SchemeGroupVersion.Group, Version: extensionsv1alpha1.SchemeGroupVersion.Version, Resource: "operatingsystemconfigs"},
+		artifact{
+			newEntity: func() client.Object { return new(extensionsv1alpha1.OperatingSystemConfig) },
+			validateResource: func(n, _ client.Object) field.ErrorList {
+				new := n.(*extensionsv1alpha1.OperatingSystemConfig)
+				return validation.ValidateOperatingSystemConfig(new)
+			},
+			validateResourceUpdate: func(n, o client.Object) field.ErrorList {
+				new := n.(*extensionsv1alpha1.OperatingSystemConfig)
+				old := o.(*extensionsv1alpha1.OperatingSystemConfig)
+				return validation.ValidateOperatingSystemConfigUpdate(new, old)
+			}})
+
+	h.addResource(metav1.GroupVersionResource{Group: extensionsv1alpha1.SchemeGroupVersion.Group, Version: extensionsv1alpha1.SchemeGroupVersion.Version, Resource: "workers"},
+		artifact{
+			newEntity: func() client.Object { return new(extensionsv1alpha1.Worker) },
+			validateResource: func(n, _ client.Object) field.ErrorList {
+				new := n.(*extensionsv1alpha1.Worker)
+				return validation.ValidateWorker(new)
+			},
+			validateResourceUpdate: func(n, o client.Object) field.ErrorList {
+				new := n.(*extensionsv1alpha1.Worker)
+				old := o.(*extensionsv1alpha1.Worker)
+				return validation.ValidateWorkerUpdate(new, old)
+			}})
+
+	return &h
+}
+
+type handler struct {
+	reader    client.Reader
+	decoder   *admission.Decoder
+	logger    logr.Logger
+	artifacts map[metav1.GroupVersionResource]artifact
+}
+
+func (h *handler) addResource(gvr metav1.GroupVersionResource, art artifact) {
+	h.artifacts[gvr] = art
+}
+
+var _ admission.Handler = &handler{}
+
+func (h *handler) InjectAPIReader(reader client.Reader) error {
+	h.reader = reader
+	return nil
+}
+
+func (h *handler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+func (h *handler) Handle(ctx context.Context, ar admission.Request) admission.Response {
+	_, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	artifact, ok := h.artifacts[ar.Resource]
+	if !ok {
+		return admission.Allowed("validation not found for the given resource")
+	}
+
+	switch ar.Operation {
+	case v1.Create:
+		return h.handleValidation(ar.Object, ar.OldObject, artifact.newEntity, artifact.validateResource)
+	case v1.Update:
+		return h.handleValidation(ar.Object, ar.OldObject, artifact.newEntity, artifact.validateResourceUpdate)
+	default:
+		return admission.Allowed("operation is not CREATE or UPDATE")
+	}
+}
+
+type newEntity func() client.Object
+type validate func(new, old client.Object) field.ErrorList
+
+// artifact servers as a helper to setup the corresponding function.
+type artifact struct {
+	// newEntity is a simple function that creates and returns a new resource.
+	newEntity newEntity
+
+	// validateResource is a wrapper function for the different create validation functions.
+	validateResource validate
```

Comment:
Isn't `validateCreateResource` a better name? So currently one has to check the doc string to understand that is for CREATE?

Reply:
IMO - yes it is! 

I was trying to be consistent with `pkg/apis/extensions/validation/...` names . If you disagree with this - please respond and I will rename it.

Reply:
I think it makes sense to rename to `validateCreateResource` to clearly point that this func is invoked when the request operation is CREATE.